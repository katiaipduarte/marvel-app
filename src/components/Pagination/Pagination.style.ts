import styled from 'styled-components';

export const PaginationContainer = styled.div`
  padding: 2rem 0;
  display: flex;
  justify-content: center;

  .separator {
    width: 1.5rem;
    margin: 0 0.25rem;
    cursor: default;
    color: ${({ theme }) => theme.primaryColor};
    font-size: 1.5rem;
    text-align: center;
  }

  .pageItem {
    background: transparent;
    border: none;
    height: 3rem;
    width: 3rem;
    margin: 0 0.25rem;
    border-radius: 50%;
    font-weight: 600;
    color: ${({ theme }) => theme.primaryColor};
    font-size: 1rem;
    cursor: pointer;
    &:hover:not(&.active) {
      background-color: ${({ theme }) => theme.primaryColor};
      color: white;
      text-decoration: none;
      box-shadow: transparent 0px 0px 0px 1px, transparent 0px 0px 0px 4px, rgba(0, 0, 0, 0.12) 0px 6px 16px;
    }
    &:focus {
      outline: 0;
    }

    &:not(.active):not(.sides):not(&:nth-last-child(2)) {
      display: none;
    }
  }

  .active {
    background-color: ${({ theme }) => theme.primaryColor};
    color: white;
    cursor: default;
    pointer-events: none;
  }

  .sides {
    box-shadow: transparent 0px 0px 0px 1px, transparent 0px 0px 0px 4px, rgba(0, 0, 0, 0.18) 0px 2px 4px;
    &:hover {
      text-decoration: none;
      box-shadow: transparent 0px 0px 0px 1px, transparent 0px 0px 0px 4px, rgba(0, 0, 0, 0.12) 0px 6px 16px;
    }
  }

  @media (min-width: 768px) {
    .pageItem {
      display: initial !important;
    }
  }
`;
