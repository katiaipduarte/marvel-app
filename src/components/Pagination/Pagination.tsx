import { PaginationContainer } from './Pagination.style';

export type Props = {
  onChangePage: (page: number) => void;
  currentPage: number;
  total: number;
};

const Pagination = (props: Props): JSX.Element => {
  const { onChangePage, currentPage, total } = props;
  const totalPages = Math.ceil(total / 20);

  const renderDots = (): JSX.Element => <div className="separator">...</div>;

  return (
    <PaginationContainer>
      {currentPage !== 1 && (
        <button
          onClick={() => onChangePage(currentPage - 1)}
          type="button"
          className="pageItem sides"
          data-testid="button"
        >
          &lt;
        </button>
      )}
      <button
        onClick={() => onChangePage(1)}
        type="button"
        className={currentPage === 1 ? 'active pageItem' : 'pageItem'}
        data-testid="button"
      >
        {1}
      </button>
      {currentPage > 3 && renderDots()}
      {currentPage === totalPages && totalPages > 3 && (
        <button onClick={() => onChangePage(currentPage - 2)} type="button" className="pageItem" data-testid="button">
          {currentPage - 2}
        </button>
      )}
      {currentPage > 2 && (
        <button onClick={() => onChangePage(currentPage - 1)} type="button" className="pageItem" data-testid="button">
          {currentPage - 1}
        </button>
      )}
      {currentPage !== 1 && currentPage !== totalPages && (
        <button
          onClick={() => onChangePage(currentPage)}
          type="button"
          className="pageItem active"
          data-testid="button"
        >
          {currentPage}
        </button>
      )}
      {currentPage < totalPages - 1 && (
        <button onClick={() => onChangePage(currentPage + 1)} type="button" className="pageItem" data-testid="button">
          {currentPage + 1}
        </button>
      )}
      {currentPage === 1 && totalPages > 3 && (
        <button onClick={() => onChangePage(currentPage + 2)} type="button" className="pageItem" data-testid="button">
          {currentPage + 2}
        </button>
      )}
      {currentPage < totalPages - 2 && renderDots()}
      <button
        onClick={() => onChangePage(totalPages)}
        type="button"
        className={currentPage === totalPages ? 'active pageItem' : 'pageItem'}
        data-testid="button"
      >
        {totalPages}
      </button>
      {currentPage !== totalPages && (
        <button
          onClick={() => onChangePage(currentPage + 1)}
          type="button"
          className="pageItem sides"
          data-testid="button"
        >
          &gt;
        </button>
      )}
    </PaginationContainer>
  );
};

export default Pagination;
