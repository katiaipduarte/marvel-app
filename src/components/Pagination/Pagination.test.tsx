import Pagination, { Props } from './Pagination';
import { render, screen, fireEvent } from '@testing-library/react';

describe('the <Pagination /> component', () => {
  it('renders the component from the beginning', () => {
    const mockProps: Props = {
      currentPage: 1,
      total: 100,
      onChangePage: jest.fn(),
    };

    const { container } = render(<Pagination {...mockProps} />);

    const buttons = screen.queryAllByTestId('button');
    fireEvent.click(buttons[0]);
    expect(mockProps.onChangePage).toBeCalledWith(1);
    fireEvent.click(buttons[1]);
    expect(mockProps.onChangePage).toBeCalledWith(2);
    fireEvent.click(buttons[2]);
    expect(mockProps.onChangePage).toBeCalledWith(3);
    fireEvent.click(buttons[3]);
    expect(mockProps.onChangePage).toBeCalledWith(5);
    fireEvent.click(buttons[4]);
    expect(mockProps.onChangePage).toBeCalledWith(2);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the component from the end', () => {
    const mockProps: Props = {
      currentPage: 5,
      total: 100,
      onChangePage: jest.fn(),
    };

    const { container } = render(<Pagination {...mockProps} />);

    const buttons = screen.queryAllByTestId('button');
    fireEvent.click(buttons[0]);
    expect(mockProps.onChangePage).toBeCalledWith(4);
    fireEvent.click(buttons[1]);
    expect(mockProps.onChangePage).toBeCalledWith(1);
    fireEvent.click(buttons[2]);
    expect(mockProps.onChangePage).toBeCalledWith(3);
    fireEvent.click(buttons[3]);
    expect(mockProps.onChangePage).toBeCalledWith(4);
    fireEvent.click(buttons[4]);
    expect(mockProps.onChangePage).toBeCalledWith(5);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders the component from the middle', () => {
    const mockProps: Props = {
      currentPage: 3,
      total: 100,
      onChangePage: jest.fn(),
    };

    const { container } = render(<Pagination {...mockProps} />);

    const buttons = screen.queryAllByTestId('button');
    fireEvent.click(buttons[0]);
    expect(mockProps.onChangePage).toBeCalledWith(2);
    fireEvent.click(buttons[1]);
    expect(mockProps.onChangePage).toBeCalledWith(1);
    fireEvent.click(buttons[2]);
    expect(mockProps.onChangePage).toBeCalledWith(2);
    fireEvent.click(buttons[3]);
    expect(mockProps.onChangePage).toBeCalledWith(3);
    fireEvent.click(buttons[4]);
    expect(mockProps.onChangePage).toBeCalledWith(4);
    fireEvent.click(buttons[5]);
    expect(mockProps.onChangePage).toBeCalledWith(5);
    fireEvent.click(buttons[6]);
    expect(mockProps.onChangePage).toBeCalledWith(4);
    expect(container.firstChild).toMatchSnapshot();
  });
});
