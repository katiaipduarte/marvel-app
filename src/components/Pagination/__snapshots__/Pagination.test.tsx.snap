// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the <Pagination /> component renders the component from the beginning 1`] = `
<div
  class="sc-bdnxRM lfNxan"
>
  <button
    class="active pageItem"
    data-testid="button"
    type="button"
  >
    1
  </button>
  <button
    class="pageItem"
    data-testid="button"
    type="button"
  >
    2
  </button>
  <button
    class="pageItem"
    data-testid="button"
    type="button"
  >
    3
  </button>
  <div
    class="separator"
  >
    ...
  </div>
  <button
    class="pageItem"
    data-testid="button"
    type="button"
  >
    5
  </button>
  <button
    class="pageItem sides"
    data-testid="button"
    type="button"
  >
    &gt;
  </button>
</div>
`;

exports[`the <Pagination /> component renders the component from the end 1`] = `
<div
  class="sc-bdnxRM lfNxan"
>
  <button
    class="pageItem sides"
    data-testid="button"
    type="button"
  >
    &lt;
  </button>
  <button
    class="pageItem"
    data-testid="button"
    type="button"
  >
    1
  </button>
  <div
    class="separator"
  >
    ...
  </div>
  <button
    class="pageItem"
    data-testid="button"
    type="button"
  >
    3
  </button>
  <button
    class="pageItem"
    data-testid="button"
    type="button"
  >
    4
  </button>
  <button
    class="active pageItem"
    data-testid="button"
    type="button"
  >
    5
  </button>
</div>
`;

exports[`the <Pagination /> component renders the component from the middle 1`] = `
<div
  class="sc-bdnxRM lfNxan"
>
  <button
    class="pageItem sides"
    data-testid="button"
    type="button"
  >
    &lt;
  </button>
  <button
    class="pageItem"
    data-testid="button"
    type="button"
  >
    1
  </button>
  <button
    class="pageItem"
    data-testid="button"
    type="button"
  >
    2
  </button>
  <button
    class="pageItem active"
    data-testid="button"
    type="button"
  >
    3
  </button>
  <button
    class="pageItem"
    data-testid="button"
    type="button"
  >
    4
  </button>
  <button
    class="pageItem"
    data-testid="button"
    type="button"
  >
    5
  </button>
  <button
    class="pageItem sides"
    data-testid="button"
    type="button"
  >
    &gt;
  </button>
</div>
`;
